src/main/java/net/snowflake/ingest/streaming/internal/StreamingIngestStage.java

```diff
@@ -225,7 +227,10 @@ static boolean isCredentialsExpiredException(Exception e) {
       return false;
     }
 
-    if (e instanceof SnowflakeSQLException) {
+    if (e instanceof SnowflakeSQLLoggedException) {
+      return ((SnowflakeSQLLoggedException) e).getErrorCode()
+          == S3_OPERATION_ERROR.getMessageCode();
+    } else if (e instanceof SnowflakeSQLException) {
```

Comment:
This is highly error-prone, the reason we need to fix it again is because the JDBC is throwing a new exception, and we have no control on whether they will update the exception type or add a new code pass that throughs a new type of exception, I think we should just catch all exceptions and always retry a few times. + @sfc-gh-lsembera to see if he has a better idea

Reply:
I dont think catching all exception is good idea here. 
I agree that this is error prone and not safe for future versions but ideally the JDBC interface should not change its response. (I think if it as a non backward compatible change)
If you check JDBC code (`handleS3Exception`), it catches all ClientException and ServerException. 


It's better we catch specific exception and work with JDBC team! 
Thoughts?

Reply:
> ideally the JDBC interface should not change its response.

Agree that catch all is not ideal, but how would you guarantee that there won't be any non-compatible change in the future? We don't have anyone who look into every JDBC PRs and make sure this won't happen. In fact, the same issues happens twice in a few months, the first fix (by Lukas) is done because they update the exception type, the second fix (this PR) is because they add a new place with a different exception

Reply:
> but how would you guarantee that there won't be any non-compatible change in the future?
I think this is something that the library owners should worry about. We as a consumer would have to go through these set of changes everytime we upgrade because a version we want to use fixes other issues. 

having that said, I dont have a strong preference of either of them, I just feel catch all is suppressing all exceptions without knowing what the root cause is.

Reply:
or more like refresh the token regardless of an exception on the final attempt?

Reply:
The benefit of doing it in the final attempt would be we would issue fewer token renewals, but on the other hand it would introduce latency because the token would only be renewed on 5th retry. I don't know which one is a bigger concern.

Reply:
+1 on always refresh the token on the first exception, and then retry 5 times if the exception persisted

Reply:
Yes, its implemented, PTAL! thanks